Nil pointer dereference while receiving message

Problem happened with version `0.3.8-0080`.
First of all, one of addresses became unavailable.
The one, which starts with `303:`.
Then, after some time, Yggdrasil server crashed.
Since logging for Windows service is disabled, I can't say exactly what happened.
Restarted it in console mode now and will wait for a crash.
If it reproduces, I will update report, if not - will close it.

Here is the crash:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x5990ad]

goroutine 1696296 [running]:
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*Conn).recvMsg(0x0, 0x6
baa80, 0xc0003c8240, 0xc0001d0000, 0x144, 0x178)
        /home/circleci/project/src/yggdrasil/conn.go:184 +0x7d
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*sessionInfo)._recvPack
et.func1.1()
        /home/circleci/project/src/yggdrasil/session.go:479 +0x18c
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*sessionInfo).checkCall
backs.func1()
        /home/circleci/project/src/yggdrasil/session.go:537 +0xcf
github.com/Arceliar/phony.(*Inbox).run(0xc0003c8240)
        /go/pkg/mod/github.com/!arceliar/phony@v0.0.0-20190831214819-9b642ea019a
d/actor.go:83 +0x12d
created by github.com/Arceliar/phony.(*Inbox).enqueue
        /go/pkg/mod/github.com/!arceliar/phony@v0.0.0-20190831214819-9b642ea019a
d/actor.go:51 +0xda
```

--- a/src/yggdrasil/conn.go
+++ b/src/yggdrasil/conn.go
@@ -90,45 +90,47 @@ func (c *Conn) setMTU(from phony.Actor, mtu uint16) {
 
 // This should never be called from the router goroutine, used in the dial functions
 func (c *Conn) search() error {
-	var sinfo *searchInfo
-	var isIn bool
+	var err error
+	done := make(chan struct{})
 	phony.Block(&c.core.router, func() {
-		sinfo, isIn = c.core.router.searches.searches[*c.nodeID]
-	})
-	if !isIn {
-		done := make(chan struct{}, 1)
-		var sess *sessionInfo
-		var err error
-		searchCompleted := func(sinfo *sessionInfo, e error) {
-			sess = sinfo
-			err = e
-			// FIXME close can be called multiple times, do a non-blocking send instead
-			select {
-			case done <- struct{}{}:
-			default:
+		_, isIn := c.core.router.searches.searches[*c.nodeID]
+		if !isIn {
+			searchCompleted := func(sinfo *sessionInfo, e error) {
+				select {
+				case <-done:
+					// Somehow this was called multiple times, TODO don't let that happen
+					if sinfo != nil {
+						// Need to clean up to avoid a session leak
+						sinfo.cancel.Cancel(nil)
+					}
+				default:
+					if sinfo != nil {
+						// Finish initializing the session
+						sinfo.conn = c
+					}
+					c.session = sinfo
+					err = e
+					close(done)
+				}
 			}
-		}
-		phony.Block(&c.core.router, func() {
-			sinfo = c.core.router.searches.newIterSearch(c.nodeID, c.nodeMask, searchCompleted)
+			sinfo := c.core.router.searches.newIterSearch(c.nodeID, c.nodeMask, searchCompleted)
 			sinfo.continueSearch()
-		})
-		<-done
-		c.session = sess
-		if c.session == nil && err == nil {
-			panic("search failed but returned no error")
+		} else {
+			err = errors.New("search already exists")
+			close(done)
 		}
-		if c.session != nil {
-			c.nodeID = crypto.GetNodeID(&c.session.theirPermPub)
-			for i := range c.nodeMask {
-				c.nodeMask[i] = 0xFF
-			}
-			c.session.conn = c
-		}
-		return err
-	} else {
-		return errors.New("search already exists")
+	})
+	<-done
+	if c.session == nil && err == nil {
+		panic("search failed but returned no error")
 	}
-	return nil
+	if c.session != nil {
+		c.nodeID = crypto.GetNodeID(&c.session.theirPermPub)
+		for i := range c.nodeMask {
+			c.nodeMask[i] = 0xFF
+		}
+	}
+	return err
 }
 
 // Used in session keep-alive traffic

